<a href="https://colab.research.google.com/github/joannedonohue/classification-challenge/blob/main/spam_detector_Final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Spam Detector

You work at an Internet Service Provider (ISP) and you've been tasked with improving the email filtering system for its customers. You've been provided with a dataset that contains information about emails, with two possible classifications: spam and not spam. The ISP wants you to take this dataset and develop a supervised machine learning (ML) model that will accurately detect spam emails so it can filter them out of its customers' inboxes.

You will be creating two classification models to fit the provided data, and evaluate which model is more accurate at detecting spam. The models you'll create will be a logistic regression model and a random forest model.

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data directly from the web location
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

,word_freq_make,word_freq_address,word_freq_all,word_freq_3d,word_freq_our,word_freq_over,word_freq_remove,word_freq_internet,word_freq_order,word_freq_mail,...,char_freq_;,char_freq_(,char_freq_[,char_freq_!,char_freq_$,char_freq_#,capital_run_length_average,capital_run_length_longest,capital_run_length_total,spam
0,0.00,0.64,0.64,0.0,0.32,0.00,0.00,0.00,0.00,0.00,...,0.00,0.000,0.0,0.778,0.000,0.000,3.756,61,278,1
1,0.21,0.28,0.50,0.0,0.14,0.28,0.21,0.07,0.00,0.94,...,0.00,0.132,0.0,0.372,0.180,0.048,5.114,101,1028,1
2,0.06,0.00,0.71,0.0,1.23,0.19,0.19,0.12,0.64,0.25,...,0.01,0.143,0.0,0.276,0.184,0.010,9.821,485,2259,1
3,0.00,0.00,0.00,0.0,0.63,0.00,0.31,0.63,0.31,0.63,...,0.00,0.137,0.0,0.137,0.000,0.000,3.537,40,191,1
4,0.00,0.00,0.00,0.0,0.63,0.00,0.31,0.63,0.31,0.63,...,0.00,0.135,0.0,0.135,0.000,0.000,3.537,40,191,1


In [3]:
# Explore the features
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4601 entries, 0 to 4600
Data columns (total 58 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   word_freq_make              4601 non-null   float64
 1   word_freq_address           4601 non-null   float64
 2   word_freq_all               4601 non-null   float64
 3   word_freq_3d                4601 non-null   float64
 4   word_freq_our               4601 non-null   float64
 5   word_freq_over              4601 non-null   float64
 6   word_freq_remove            4601 non-null   float64
 7   word_freq_internet          4601 non-null   float64
 8   word_freq_order             4601 non-null   float64
 9   word_freq_mail              4601 non-null   float64
 10  word_freq_receive           4601 non-null   float64
 11  word_freq_will              4601 non-null   float64
 12  word_freq_people            4601 non-null   float64
 13  word_freq_report            4601 

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct!

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

## Prediction:

The Random Forest model is more likely to perform better than Logistic Regression due to the ability to capture non-linear patterns and handle complex data, like text and features of an email.

## Split the Data into Training and Testing Sets

In [4]:
# Create the labels set `y` and features DataFrame `X`
# The data is labeled with SPAM 0 or 1 so we will use this to train our prediction model
y = data["spam"]
X = data.drop(columns="spam")

In [5]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

,count
spam,
0,2788
1,1813


# Observation:

* 61% of emails are not spam (0)
* The remaining is flagged as spam (1)
* This is a good mix of 0 and 1 data to build our model, but could be more balanced

In [9]:
# Split the data into X_train, X_test, y_train, y_test
# Setting random state to 1 for repeatable results
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

# Review train X sample:
X_train.head()

,word_freq_make,word_freq_address,word_freq_all,word_freq_3d,word_freq_our,word_freq_over,word_freq_remove,word_freq_internet,word_freq_order,word_freq_mail,...,word_freq_conference,char_freq_;,char_freq_(,char_freq_[,char_freq_!,char_freq_$,char_freq_#,capital_run_length_average,capital_run_length_longest,capital_run_length_total
4576,0.00,0.0,0.00,0.0,0.00,0.0,0.0,0.0,0.0,0.0,...,0.0,0.000,0.131,0.000,0.0,0.000,0.0,1.488,5,64
4401,0.00,0.0,0.00,0.0,0.00,0.0,0.0,0.0,0.0,0.0,...,0.0,0.000,0.000,0.000,0.0,0.000,0.0,1.571,5,11
3707,0.17,0.0,0.17,0.0,0.00,0.0,0.0,0.0,0.8,0.0,...,0.0,0.253,0.168,0.084,0.0,0.024,0.0,4.665,81,1031
2362,0.00,0.0,0.00,0.0,0.00,0.0,0.0,0.0,0.0,0.0,...,0.0,0.000,0.000,0.000,0.0,0.000,0.0,4.228,53,148
1537,0.00,0.0,0.00,0.0,2.17,0.0,0.0,0.0,0.0,0.0,...,0.0,0.000,0.000,0.000,0.0,0.000,0.0,1.333,5,16


## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [14]:
from sklearn.preprocessing import StandardScaler

# Scaling data to ensure all features are on the same scale

# Create the StandardScaler instance
scaler = StandardScaler()

In [15]:
# Fit the Standard Scaler with the training data
# Calculates the mean and std dev using only the training data
scaler.fit(X_train)


StandardScaler()

In [18]:
# Scale the training data and test using the same mean and std dev from train data
X_test_scaled = scaler.transform(X_test)
X_train_scaled = scaler.transform(X_train)

# Check Results
#print(X_test_scaled)
#print(X_train_scaled)

# Create DataFrames with the scaled data
df_X_test_scaled = pd.DataFrame(X_test_scaled, columns=X_test.columns)
df_X_train_scaled = pd.DataFrame(X_train_scaled, columns=X_train.columns)

# Check Results
print(df_X_test_scaled.head())
print(df_X_train_scaled.head())

   word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0       -0.358119          -0.167442      -0.560720     -0.044647   
1       -0.358119          -0.167442      -0.560720     -0.044647   
2       -0.358119           0.243092      -0.560720     -0.044647   
3       -0.358119          -0.167442      -0.105039     -0.044647   
4       -0.358119          -0.167442      -0.560720     -0.044647   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0       1.040797       -0.349536         -0.289221            -0.25739   
1      -0.459893       -0.349536         -0.289221            -0.25739   
2       1.115089       -0.349536          3.894261            -0.25739   
3       0.907072       -0.349536         -0.289221            -0.25739   
4      -0.459893       -0.349536         -0.289221            -0.25739   

   word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
0        -0.325494       -0.368236  ...             -0.105122 

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like.

In [34]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

logistic_regression_model = LogisticRegression(random_state=1, max_iter=120)

lr_model = logistic_regression_model.fit(df_X_train_scaled, y_train)

print(f"Training Data Score: {lr_model.score(df_X_train_scaled, y_train):.2%}")
print(f"Testing Data Score: {lr_model.score(df_X_test_scaled, y_test):.2%}")

Training Data Score: 92.96%
Testing Data Score: 92.79%


In [41]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions_lr = lr_model.predict(df_X_test_scaled)

# Review the predictions
testing_predictions_lr

array([0, 0, 1, ..., 0, 0, 1])

In [42]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
print(f"Accuracy Score: {accuracy_score(y_test, testing_predictions_lr):.2%}")

Accuracy Score: 92.79%


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like.

In [49]:
from sklearn.model_selection import GridSearchCV

# Using GridSearch to optimize parameters

param_grid = {
    'n_estimators': [64, 128, 256],
    'max_depth': [10, 20, None],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}
grid_search = GridSearchCV(RandomForestClassifier(random_state=1), param_grid, cv=5)
grid_search.fit(df_X_train_scaled, y_train)
print(f"Best parameters: {grid_search.best_params_}")

Best parameters: {'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 256}


In [52]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier

random_forest_model = RandomForestClassifier(n_estimators = 256, random_state = 1)

rf_model = random_forest_model.fit(df_X_train_scaled, y_train)

print(f"Training Data Score: {rf_model.score(df_X_train_scaled, y_train):.2%}")
print(f"Testing Data Score: {rf_model.score(df_X_test_scaled, y_test):.2%}")

Training Data Score: 99.97%
Testing Data Score: 96.52%


In [53]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions_rf = rf_model.predict(df_X_test_scaled)

# Review the predictions
testing_predictions_rf

array([1, 1, 1, ..., 1, 0, 1])

In [54]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
print(f"Accuracy Score: {accuracy_score(y_test, testing_predictions_rf):.2%}")

Accuracy Score: 96.52%


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

## Confusion matrix to further evaluate the models

In [55]:
# Create confusion matrix to gain insights into model performance
from sklearn.metrics import classification_report, confusion_matrix

# Logistic Regression metrics
print("Logistic Regression - Classification Report")
print(classification_report(y_test, testing_predictions_lr))
print("Confusion Matrix")
print(confusion_matrix(y_test, testing_predictions_lr))

# Random Forest metrics
print("\nRandom Forest - Classification Report")
print(classification_report(y_test, testing_predictions_rf))
print("Confusion Matrix")
print(confusion_matrix(y_test, testing_predictions_rf))


Logistic Regression - Classification Report
              precision    recall  f1-score   support

           0       0.92      0.96      0.94       701
           1       0.94      0.87      0.90       450

    accuracy                           0.93      1151
   macro avg       0.93      0.92      0.92      1151
weighted avg       0.93      0.93      0.93      1151

Confusion Matrix
[[676  25]
 [ 58 392]]

Random Forest - Classification Report
              precision    recall  f1-score   support

           0       0.96      0.98      0.97       701
           1       0.97      0.94      0.95       450

    accuracy                           0.97      1151
   macro avg       0.97      0.96      0.96      1151
weighted avg       0.97      0.97      0.97      1151

Confusion Matrix
[[687  14]
 [ 26 424]]


## Model Evaluation Results

### Which model performed better?
The Random Forest model performed better than the Logistic Regression model across all key metrics.

- **Accuracy**: Random Forest achieved a higher accuracy of 97% compared to 93% for Logistic Regression.
- **Precision and Recall**: Random Forest had better precision (97%) and recall (94%) for detecting spam emails, compared to Logistic Regression's precision (94%) and recall (87%).
- **F1-Score**: The Random Forest model had a higher F1-score, indicating a better balance between precision and recall.

### How does this compare to your prediction?
This result aligns with my initial prediction that the Random Forest model would outperform Logistic Regression. The Random Forest model’s ability to handle non-linear relationships and capture complex patterns in the data contributed to its performance.

Logistic Regression, while simpler and faster, struggled to match the performance of the more robust Random Forest model.
